### ISSUE TITLE
Style rule: val constants can be declared as const

### ISSUE THREAD

**Issue by mkobit** [arturbosch/detekt#731]
IDEA introduced this inspection in 1.2.20 ([changelog](https://github.com/JetBrains/kotlin/blob/master/ChangeLog.md#new-features-5)) in https://youtrack.jetbrains.com/issue/KT-20644

The description of the quickfix:

> This inspection reports object and top-level val that might be declared as const for better performance and Java interoperability.

Here are a few examples and a counter example:

```kotlin
val topLevel = "TopLevel"

class MyClass {
    companion object {
        val compValue = "Companion"
    }
}

object MyObject {
    val compValue = "Companion"
}

fun myFun() {
    // const not applicable here, so should not be detected
    val localValue = "Local"
}
```

**Comment by Mauin** [comment]
As mentioned in the PR for this rule. The new `MayBeConst` rule does identify most of the basic cases in which properties can be `const val`. However there are more cases where `val`s are initialized with expressions that could allow for `const val`s instead. But without type resolution this won't be possible for now.
We can and should revisit this rule once type resolution is available.
